src/components/ModelingSidebar/ModelingPanes/CodeEditor.tsx

```diff
@@ -0,0 +1,140 @@
+import React, { useEffect, useMemo, useRef, useState } from 'react'
+import {
+  EditorState,
+  EditorStateConfig,
+  Extension,
+  StateEffect,
+} from '@codemirror/state'
+import { EditorView } from '@codemirror/view'
+import { oneDark } from '@codemirror/theme-one-dark'
+
+//reference: https://github.com/sachinraja/rodemirror/blob/main/src/use-first-render.ts
+const useFirstRender = () => {
+  const firstRender = useRef(true)
+
+  useEffect(() => {
+    firstRender.current = false
+  }, [])
+
+  return firstRender.current
+}
+
+const defaultLightThemeOption = EditorView.theme(
+  {
+    '&': {
+      backgroundColor: '#fff',
+    },
+  },
+  {
+    dark: false,
+  }
+)
+
+interface ICodeEditor {
+  onView?: (view: EditorView | null) => void
+  initialDocValue?: EditorStateConfig['doc']
+  extensions?: Extension
+  theme: 'light' | 'dark'
+  autoFocus?: boolean
+  selection?: EditorStateConfig['selection']
+}
+
+interface UseCodeMirror extends ICodeEditor {
+  container?: HTMLDivElement | null
+}
+
+const CodeEditor: React.FC<ICodeEditor> = ({
+  onView,
+  extensions = [],
+  initialDocValue,
+  theme,
+  autoFocus = false,
+  selection,
+}) => {
+  const editorRef = useRef<HTMLDivElement>(null)
+
+  useCodeMirror({
+    container: editorRef.current,
+    onView,
+    extensions,
+    initialDocValue,
+    theme,
+    autoFocus,
+    selection,
+  })
+
+  return <section ref={editorRef}></section>
+}
+
+export function useCodeMirror(props: UseCodeMirror) {
+  const {
+    onView,
+    extensions = [],
+    initialDocValue,
+    theme,
+    autoFocus = false,
+    selection,
+  } = props
+
+  const [container, setContainer] = useState<HTMLDivElement | null>()
+  const [editorView, setEditorView] = useState<EditorView | null>(null)
+
+  const isFirstRender = useFirstRender()
```

Comment:
nit: could've just done useState(false) and have a useEffect instead of a new function and using a useRef but effectively they equate to the same thing

Reply:
Sorry I think I didn't clearly convey how bad at react I am, this is mostly copy pasta so its like you are saying words I just don't know what they mean 😭

Reply:
I think the useFirstRender is okay/better, since it's using a useRef, which means it can be changed without triggering another render, I'm guessing that was the intent.